### PR TITLE
Convert old rest stack: supremm module

### PR DESCRIPTION
**Pull request by jsperhac**
This is a step along the way toward killing off the old REST stack.

## Description
Create a NewRest-style controller for the internal dashboard's supremm dataflow functionality, to replace the old classes/REST/supremm code. 

Refer to pull requests against 
- xdmod (https://github.com/ubccr/xdmod/pull/73), that: removes entirely-unused source code from the old classes/REST rest stack, and removes associated documentation, html/rest_catalog/ 
- xdmod-appkenel (https://github.com/ubccr/xdmod-appkernels/pull/12), that converts another old REST stack remnant to use the NewRest controllers.
- old pull request #422 from the xdmod-private ccr repo, which did not go quite this far in clearing unused code from the old REST stack

## Motivation and Context
- Head in the direction of reducing 3 rest stacks to 2 rest stacks
- I'm tired of duplicate code.

## Tests performed
Changes are limited to the internal dashboard's supremm dataflow, which reports on the last time various supremm sources contributed data. These were tested by hand to verify, using chrome debugger to observe which rest endpoints are in use. Finally, tests for the new endpoints have been added and use the integration test harness.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Caveats:
More things I'd like to do related to the rest stacks (but not a part of the current pull request):
- Introduce improved logging for the NewRest controllers

## To Do:
Remaining to do in this exercise:
- see pull requests named "convert-old-rest-stack" for appkernel and base xdmod modules. 
- following merging of the convert-old-rest-stack pull requests, delete entire contents of REST/, and review the contents of libraries/rest.php. This will be addressed in a subsequent pull request.
- Move this and other iframes in the internal dashboard into proper tabs. (And: Review ~duplicated code such as internal_dashboard/user_check.php and internal_dashboard/supremm/user_check.php.) This should be addressed in a separate pull request.

### CODE REVIEW

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+                $payload['message'] = self::_DEFAULT_ERROR_MESSAGE;
+            }
+
+        } catch (BadRequestHttpException $e) {
+            $payload['message'] = $e->getMessage();
+            $statusCode = $e->getStatusCode();
+        } catch (HttpException $e) {
+            $payload['message'] = $e->getMessage();
+            $statusCode = $e->getStatusCode();
+        } catch (\Exception $e) {
```
Comment (jpwhite4) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: If authorize() throws an AccessDeniedException then this code will catch it and change the status code to internal error. This is wrong.

I think that you should not be catching exceptions here. Instead let the framework handle them

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        return null;
+    } // function _queryDbstats
+
+    /**
+     * Format and label timestamp appropriately for report output (minutes, hours, weeks, or months).
+     *
+     * @param timestamp $ts
+     * @return mixed
+     */
+    private function _time2str($ts)
```
Comment (plessbd) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: I feel like this should be a utility function that anything can use.  Maybe even remove this from the backend code and use something like moment in the frontend to do this as a display thing...

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        }
+    } // function _time2str()
+
+    /**
+     * Format and label data sizes appropriately for report output (as GB, MB, kB, or B).
+     *
+     * @param numeric $d
+     * @return mixed
+     */
+    private function _formatDataSize($d)
```
Comment (plessbd) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: This as well seems like a utility function that is able to be used in multiple places.

**Comment by plessbd** [comment]
I think it would help to run the code through phpcbf to fix a bunch of the warnings and errors from travis ci.

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+        // QUERY ROUTES
+        $controller
+            ->get("$root/resources", "$base::getResources");
+
+        $controller
+            ->get("$root/dbstats", "$base::getDbstats")
+            ->assert('resource_id', '(\w|_|-])+')
+            ->convert('db_id', "$conversions::toInt")
+            ->convert('resource_id', "$conversions::toString");
```
Comment (jsperhac) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: Yup, it's wrong here, I will fix. Maybe changing the variable name will also help clarify this.

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+                } else {
+
+                    $payload['message'] = "There was no result found for the given database ($dbid) and resource ($resourceid)";
+                    $statusCode = 404;
+                }
+            } else {
+                $payload['message'] = self::_DEFAULT_ERROR_MESSAGE;
+            }
+        } catch (BadRequestHttpException $e) {
```
Comment (jsperhac) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: This is modeled rather heavily on the other NewRest controllers, which similarly catch lots of exceptions. I will review...

**Comment by jsperhac** [comment]
And I missed that travis ci failed. OK, I'll revisit. Thanks.

File: html/internal_dashboard/supremm/stats.php [hunk 6]
```diff
@@ -1,17 +1,27 @@
...
         <script type="text/javascript" src="/gui/lib/jsPlumb/jquery.jsPlumb-1.6.3-min.js"></script>
-        <script>
-<?php echo "var token='" . $_SESSION['session_token'] . "';" ; ?>
+
+        <!-- Ext and Jquery adapter -->
+        <script type="text/javascript" src="/gui/lib/extjs/adapter/jquery/ext-jquery-adapter.js"></script>
+        <script type="text/javascript" src="/gui/lib/extjs/ext-all-debug.js"></script>
+
+        <!-- RESTProxy -->
+        <script type="text/javascript" src="../../gui/js/CCR.js"></script>
```
Comment (jpwhite4) [inline html/internal_dashboard/supremm/stats.php]: Why are you including the REST.js RESTProxy and CCR.js javascript files? They are not needed.

Comment (jsperhac) [inline html/internal_dashboard/supremm/stats.php]: I believe you are right...duly noted, I'll remove those includes.

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php (hunk 2 above)
Comment (jsperhac) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: For the moment the idea is to port what's there. Moving general functionality into utility functions should absolutely be done separately...

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php (hunk 3 above)
Comment (jsperhac) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: As mentioned above, yes, I agree--in a separate pull request.

File: classes/NewRest/Controllers/SupremmDataflowControllerProvider.php (hunk 5 above)
Comment (jsperhac) [inline classes/NewRest/Controllers/SupremmDataflowControllerProvider.php]: Exception code has been removed; framework handles this, as you point out!

File: html/internal_dashboard/supremm/stats.php (hunk 6 above)
Comment (jsperhac) [inline html/internal_dashboard/supremm/stats.php]: Done.

**Comment by jsperhac** [comment]
In order for integration tests to work, the fix-framework-authorize-exceptions branch on xdmod repo will be needed (pull request to follow once unit tests exist)

**Comment by plessbd** [comment]
Existing travis CI Errors will be fixed by #35 and an update to https://github.com/jpwhite4/lint-diff

**Pull request merged**